core/src/main/java/com/linecorp/armeria/client/endpoint/SlowStartAwareEndpointSelectionStrategy.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.endpoint;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.linecorp.armeria.client.ClientFactory;
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.client.endpoint.WeightedRoundRobinStrategy.EndpointsAndWeights;
+
+import io.netty.util.concurrent.ScheduledFuture;
+
+/**
+ * A weighted round-robin strategy that lamp up the weight gradually from zero to its normal weight value.
+ */
+public final class SlowStartAwareEndpointSelectionStrategy implements EndpointSelectionStrategy {
+    private final EndpointWeighter endpointWeighter;
+    private final ClientFactory clientFactory;
+    private final Duration slowStartInterval;
+    private final int numberOfSteps;
+
+    /**
+     * Creates a {@link SlowStartAwareEndpointSelectionStrategy}.
+     */
+    public SlowStartAwareEndpointSelectionStrategy(EndpointWeighter endpointWeighter,
+                                                   ClientFactory clientFactory,
+                                                   Duration slowStartInterval,
+                                                   int numberOfSteps) {
+        this.endpointWeighter = endpointWeighter;
+        this.clientFactory = clientFactory;
+        this.slowStartInterval = requireNonNull(slowStartInterval, "slowStartInterval");
+        checkArgument(numberOfSteps > 0, "numberOfSteps(%s) > 0");
+        this.numberOfSteps = numberOfSteps;
+    }
+
+    @Override
+    public SlowStartAwareWeightedRoundRobinSelector newSelector(EndpointGroup endpointGroup) {
```

Comment:
Could return `EndpointSelector` because `SlowStartAwareWeightedRoundRobinSelector` is private.